### DISCUSSION TITLE
When publish into IIS a html resource it is not available with 404 error

### DISCUSSION THREAD

**Discussion by kudryavtsevda** [OrchardCMS/OrchardCore#11362] [Q&A]
I have GrapesJS-editor.html resource under wwwroot folder in module named OrchardCoreContrib.Html. 
When I run the app under IIS express or via dotnet run the resource is available and no error.

But when I try to publish it into IIS via `dotnet publish -c debug` I get 404 error and this resource is not available. See screenshot below:
![image](https://user-images.githubusercontent.com/4447809/157927564-55786b41-9226-4f0d-8728-fc3ef02f3c44.png)

it is how I use it inside cshtml file:
![image](https://user-images.githubusercontent.com/4447809/157927846-daa3f9b9-23aa-4965-a440-a4cce53ddf8e.png)

location in the project:
![image](https://user-images.githubusercontent.com/4447809/157928243-c8cea7c2-9123-41da-a0d2-7db1f9f9590c.png)

It is successfull run under IIS Express from VS2022
![image](https://user-images.githubusercontent.com/4447809/157928712-d2457f42-987c-459d-bb66-7a2b6fe9084d.png)

What am I doing wrong? How to publish it into IIS with the same result as under VS2022?

**Comment by Skrypt** [comment]
Remove the leading "slash" as you are defining your src as an absolute path while it should be a relative path.

**Reply by kudryavtsevda** [reply]
> Remove the leading "slash" as you are defining your src as an absolute path while it should be a relative path.
I removed it 
![image](https://user-images.githubusercontent.com/4447809/157936363-b25907eb-4c61-49ce-ac54-b914c81a32b2.png)

And I got this result 404 bad response under VS2022 run
![image](https://user-images.githubusercontent.com/4447809/157936314-261d9802-d0d3-429f-a968-4964eafcb18c.png)

**Reply by hishamco** [reply]
Seems the assets not published. Can you confirm that the editor page inside your published folder?

**Reply by kudryavtsevda** [reply]
> Seems the assets not published. Can you confirm that the editor page inside your published folder?

Published folder doesn't contain any grapesjs-editor.html file

**Reply by hishamco** [reply]
Inside `wwwroot`?

**Reply by kudryavtsevda** [reply]
> Inside `wwwroot`?

![image](https://user-images.githubusercontent.com/4447809/157953181-1a6ccd1f-556c-413c-b0d8-b0773857c9fd.png)
yes, it is empty here

**Reply by Skrypt** [reply]
Fix the issue first with the Module manifest and maybe it will work better.

**Reply by jtkech** [reply]
Just as additional info

Module assets are not physically published but embedded in module assemblies
Also here better to use the tilde char `<iframe src="~/..."` so that it is virtual folder and tenant aware.

**Reply by Skrypt** [reply]
src here won't be processed. Would work with something like asp-src but I have a doubt here.

**Reply by kudryavtsevda** [reply]
> Just as additional info
> 
> Module assets are not physically published but embedded in module assemblies Also here better to use the tilde char `<iframe src="~/..."` so that it is virtual folder and tenant aware.

@Skrypt @jtkech 
Now module is fixed but published web-app doesn't see html resource
![image](https://user-images.githubusercontent.com/4447809/157964591-f8bcb978-e8b1-48f9-808a-52828b2d9350.png)
here my view which is running successfully
![image](https://user-images.githubusercontent.com/4447809/157965055-9307fe2c-9da4-4dac-aeff-aca9f3b2bcbf.png)

I think it is related somehow to publish and devops stuff but I am not sure

**Reply by Skrypt** [reply]
So the Controller route is working but the embedded .html file cannot be found. Make sure it is included in the .csproj. And of course it won't be in the wwwroot folder because it will be embedded in your .dll of your grapejs.module assembly. So, you need to access it by using mvc routing conventions. Normally AreaName/yourfile.html.

And what @jtkech says could also work.

**Reply by jtkech** [reply]
@Skrypt I tried it under a virtual folder, looks like that the src attribute is recognized and the tilde character is processed

@kudryavtsevda assets are not published but embedded in assemblies, can you check under VStudio that your html file is still marked as an embedded resource, and re-use the trailing `/` or maybe `~/`

**Reply by kudryavtsevda** [reply]
> So the Controller route is working but the embedded .html file cannot be found. Make sure it is included in the .csproj. And of course it won't be in the wwwroot folder because it will be embedded in your .dll of your grapejs.module assembly. So, you need to access it by using mvc routing conventions. Normally AreaName/yourfile.html.
> 
> And what @jtkech says could also work.

it is embedded but it's location under wwwroot. 


![image](https://user-images.githubusercontent.com/4447809/157970473-02e1c928-5a7f-4b91-ab22-cf610d7cdeb2.png)

![image](https://user-images.githubusercontent.com/4447809/157970777-fc2f93f7-d989-4a13-8c5a-a70e97f0f05c.png)

Should I create separate folder like "Htmls" and move html file into that folder?

**Reply by Skrypt** [reply]
It should work unless .html is not an accepted mime/type on the server. Which I would be really surprised of.

**Reply by kudryavtsevda** [reply]
> It should work unless .html is not an accepted mime/type on the server. Which I would be really surprised of.

I created folder "assets" and moved html resource into that folder. still 404 
![image](https://user-images.githubusercontent.com/4447809/157973499-13e81bdf-604a-4d88-b289-3fe767a09125.png)

![image](https://user-images.githubusercontent.com/4447809/157973629-67dd5f13-095b-4ae4-93e3-b6df202d74ae.png)

**Reply by Skrypt** [reply]
Everything contained in the Assets folder will be excluded from publishing so it is not going to work.

**Reply by kudryavtsevda** [reply]
> Everything contained in the Assets folder will be excluded from publishing so it is not going to work.

I created htmls folder and moved html resource as embeded resource. Still 404

![image](https://user-images.githubusercontent.com/4447809/157974874-70bf8407-61a6-4300-b188-1d835556579c.png)
![image](https://user-images.githubusercontent.com/4447809/157974894-faf7d002-68e7-4d3c-8b40-6826b69f7fd4.png)

**Reply by kudryavtsevda** [reply]
> Everything contained in the Assets folder will be excluded from publishing so it is not going to work.

when I marked html resource as content with copy always mode and publish I got such result. I don't understand what is going on
![image](https://user-images.githubusercontent.com/4447809/157975389-019c7c06-1b79-4cde-bcd1-d4485228909a.png)

**Reply by jtkech** [reply]
@kudryavtsevda 

First better to use `~/` with your script tag helpers.

Yes, all under the assets folder are automatically excluded from embedded resources, there are intended to be outputed by the gulp pipeline to wwwroot, if it is not configured you have to put your file under wwwroot, we embedd files under wwwroot.

But in your screen shot I can see that you removed it from embedded resource in your project file, so remove all the extra item group in your project file

**Reply by kudryavtsevda** [reply]
> @kudryavtsevda
> 
> First better to use `~/` with your script tag helpers.
> 
> Yes, all under the assets folder are automatically excluded from embedded resources, there are intended to be outputed by the gulp pipeline to wwwroot, if it is not configured you have to put your file under wwwroot, we embedd files under wwwroot.
> 
> But in your screen shot I can see that you removed it from embedded resource in your project file, so remove all the extra item group in your project file

I did as you told be. I put b.html file as embeded resource (no copy) under wwwroot and I get 404 error
![image](https://user-images.githubusercontent.com/4447809/157978695-26da0fb3-f42c-4e47-8d61-c3a07832393f.png)
this is my csproj
![image](https://user-images.githubusercontent.com/4447809/157979211-51bfb58b-b3a4-45f8-8274-df3c8ead7b84.png)

~/grapesjs.module/b.html and ~/grapesjs.module/htmls/a.html are not working

here example if ~/grapesjs.module/htmls/a.html
![image](https://user-images.githubusercontent.com/4447809/157979603-22e1b2ab-4470-436b-b139-0ef485b5ec96.png)

**Reply by jtkech** [reply]
This not exactly what I said

Remove all item groups related to folder and files, all is automatically done for you, in place of adding more

 In your paths use the exact name of your module, it is case sensitive

**Reply by kudryavtsevda** [reply]
> This not exactly what I said
> 
> Remove all item groups related to folder and files, all is automatically done for you, in place of adding more
> 
> In your paths use the exact name of your module, it is case sensitive

it is full exact path with upper B
![image](https://user-images.githubusercontent.com/4447809/157982729-df63dcb0-624e-4bcd-b071-005e50b52d28.png)

it is not exact path with lower b
![image](https://user-images.githubusercontent.com/4447809/157983096-cedc1757-da47-4de6-9e1f-659d0f6fc2fc.png)

this is csproj file
![image](https://user-images.githubusercontent.com/4447809/157983227-4dcf8d9b-cab0-4bc4-8f8c-abf8d7815068.png)

you are right! thank you a lot. I don't know how I would guess without your adivice. I much appreciate you for your paitence. 

BTW could you point me out where it is explained in tutorial or in source code?

**Reply by jtkech** [reply]
Yes you're right, for module assets the whole path is case sensitive because it is file based and to be compatible with any os, I don't think that this aspect is documented.

**Reply by jtkech** [reply]
What we are used to do is to name all our assets only with lower case letters, then we only have to take care about the modules/features names
